crates/core/src/table/parquet_config.rs

```diff
@@ -0,0 +1,166 @@
+                value: None,
+                description,
+            })
+        }
+    }
+
+    let mut v = Visitor(vec![]);
+    table_parquet_options.visit(&mut v, "", "");
+    
+    // Work around datafusion bug in file_encryption.visit.
+    // Visitor skips column_encryption_properties
+    let mut c = Visitor(vec![]);
+    if let Some(file_encryption) = &table_parquet_options.crypto.file_encryption {
+        file_encryption.column_encryption_properties.visit(&mut c, ".crypto.file_encryption", "");
+        v.0.append(&mut c.0);
+    };
+    
+    v.0
+}
+
+impl From<TableParquetOptions> for ParquetConfig {
+    fn from(table_parquet_options: TableParquetOptions) -> Self {
+        let dpo: HashMap<String, String> = Self::map_table_entries(&TableParquetOptions::default());
+        let tpo: HashMap<String, String> = Self::map_table_entries(&table_parquet_options);
+        
+        // println!("[parquet_config] Default options: {:#?}", dpo);
+        // println!("[parquet_config] Table options: {:#?}", tpo);
+        let mut altered: HashMap<String, String> = HashMap::new();
+
+        // Find altered entries
+        for (k, v_tpo) in tpo {
+            match dpo.get(&k) {
+                Some(v_dpo) => {
+                    if *v_dpo != v_tpo {
+                        altered.insert(k.clone(), v_tpo.clone());
+                    }
+                }
+                None => {altered.insert(k.clone(), v_tpo.clone()); }
+            }
+        }
+        ParquetConfig{
+            table_parquet_options: altered,
+        }
+    }
+}
+
+impl TryInto<TableParquetOptions> for ParquetConfig {
+    type Error = DataFusionError;
+    fn try_into(self) -> Result<TableParquetOptions, Self::Error> {
+        let mut options = TableParquetOptions::default();
+        for (key, value) in self.table_parquet_options {
+            if key.len() < 1 || value.len() < 1 {
+                continue;
+            }
+            // println!("Setting key: {}, value: {}", key, value);
+            options.set(&*key, &*value)?;
+        }
+        Ok(options)
+    }
+}
+
+
+impl ParquetConfig {
+    /// Create a new ParquetConfig with default options.
+    pub fn new() -> Self {
+        ParquetConfig::default()
+    }
+
+    // Convert the ParquetConfig to parquet::file::properties::WriterProperties
+    pub fn writer_options(&self) -> WriterProperties {
+
+        let mut table_parquet_options: TableParquetOptions = self.clone().try_into()
+            .expect("Failed to convert ParquetConfig to TableParquetOptions");
+        
+        table_parquet_options.global.skip_arrow_metadata = true;
+
+        // Convert TableParquetOptions to ParquetWriterOptions
+        let writer_options : ParquetWriterOptions =
+            ParquetWriterOptions::try_from(&table_parquet_options)
+                .expect("Failed to convert TableParquetOptions to ParquetWriterOptions");
+
+        writer_options.writer_options().clone()
+    }
```

Comment:
This writer_options function is essentially a kind of factory and where I would imagine hooking in a KMS (possibly with different arguments - or maybe this returns a factory?).

Reply:
Ah yeah I made a similar comment up on `crates/core/src/operations/mod.rs`. I think we might want a factory here, or somehow defer the creation of `WriterProperties` until we know the file schema and path.